src/dev/dev-server.ts

```diff
@@ -37,10 +37,22 @@ export const startDevServer = async (options: StartDevServerOptions) => {
 
   const stderr = options.stderr ?? process.stderr
 
+  const buildEvents = new EventEmitter()
+  // Allows us to block requests if we're currently rebuilding to avoid using a stale build
+  let isBuilding = false
+
+  // Used to avoid a race condition where the server attempts to process a request before the first build is complete
+  const firstBuildPromise = once(buildEvents, "built")
+
   const server = createServer(
     transformToNodeBuilder({
       defaultOrigin: `http://localhost:${options.port}`,
     })(async (req) => {
+      await firstBuildPromise
+      if (isBuilding) {
+        await once(buildEvents, "built")
+      }
```

Comment:
possible race condition if built gets sent after the if check but before the await. probably not very likely, but it will be a bitch and a half if it does happen, since that will hang until the next build.

solution:

```ts
await new Promise((res) => {
  buildEvents.on("built", res)
  if (!isBuilding) res()
})
```

Reply:
69e7eed